NIFI-1942 Processor to validate CSV against user-supplied schema

pvillard31 opened the pull request:
This processor is designed to validate a CSV formatted FlowFile against a user-supplied schema.

It leverages Cell Processors from super-csv library and gives the following options to define the expected schema:

- ParseBigDecimal
- ParseBool
- ParseChar
- ParseDate
- ParseDouble
- ParseInt
- ParseLong
- Optional
- DMinMax
- Equals
- ForbidSubStr
- LMinMax
- NotNull
- Null
- RequireHashCode
- RequireSubStr
- Strlen
- StrMinMax
- StrNotNullOrEmpty
- StrRegEx
- Unique
- UniqueHashCode

Nested cell processors are not supported except with Optional.

JPercivall commented:
Hey @pvillard31, looks like this has merge conflicts. Could you address them?

pvillard31 commented:
Hey @JPercivall, that should be OK now, thanks!

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,408 @@
+@SupportsBatching
+@InputRequirement(Requirement.INPUT_REQUIRED)
+@Tags({"csv", "schema", "validation"})
+@CapabilityDescription("Validates the contents of FlowFiles against a user-specified CSV schema")
```
Is there some sort of example CSV schema or other documentation we can point to here? I know nothing about CSV validation and if I was given the task to add/configure this processor I would be completely lost.

JPercivall replied:
Ah duh, I should have just read the additional docs portion. A comment here reminding users to checkout the additional details would suffice

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,408 @@
+        final CsvPreference csvPref = this.preference.get();
+        final boolean header = context.getProperty(HEADER).asBoolean();
+        final ComponentLog logger = getLogger();
+        final CellProcessor[] cellProcs = this.processors.get().toArray(new CellProcessor[this.processors.get().size()]);
```
Can this be done in the onScheduled? I feel there is decent amount of allocating and processing that could be re-used.

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,408 @@
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) {
+        final List<FlowFile> flowFiles = session.get(50);
```
I would suggest not doing an internal batching here. Some older processors did this because "@SupportsBatching" wasn't created yet. You already have this annotation so this will allow the user to set the batching duration instead of hiding inside the implementation.

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,408 @@
+                    "the cell processors to apply. The following cell processors are allowed in the schema definition: " +
+                    allowedOperators.toString() + ". Note: cell processors cannot be nested except with Optional.")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
```
Is there any way to do some validation here? I can see there being some pretty complicated schemas and a missing comma could lead to a lot of user headaches without any validation.

JPercivall replied:
-_- I really need to double check the commit before commenting, I see you already put this validation in the custom validate

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,408 @@
+            .displayName("Delimiter character")
+            .description("Character used as 'delimiter' in the incoming data. Example: ;")
+            .required(true)
+            .defaultValue(";")
```
Shouldn't the default delimiter for a "Comma Separated Values" file be a comma?

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,408 @@
+    private CellProcessor getProcessor(String method, String argument) {
+        switch (method) {
+
+        case "optional":
```
nit pick: the cases should be indented a level from the switch statement

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,408 @@
+            return new ParseDate(argument.substring(1, argument.length() - 1));
+
+        case "parsedouble":
+            return new ParseDouble();
```
Since this method is being used to validate, all of these no-param methods should validate that there are no arguments being passed to them.

JPercivall commented:
@pvillard31 I left a couple comments. Do you by chance have a template and/or example csv data I can use to validate the processor?

pvillard31 commented:
Hey @JPercivall, thanks for the review! I believe I addressed all of your comments. I also added the new dependency in NOTICE files (let me know if it is OK). To validate the processor I used a simple flow to generate CSV data with a ReplaceText processor: I'm gonna add the template to the JIRA right now.

JPercivall commented on nifi-assembly/NOTICE:
```diff
@@ -193,6 +193,8 @@ The following binary components are provided under the Apache Software License v
 
   (ASLv2) opencsv (net.sf.opencsv:opencsv:2.3)
 
+  (ASLv2) Super CSV (net.sf.supercsv:super-csv:2.4.0)
```
This is a Apache 2.0 licensed import with not NOTICE, it can be used with adding anything to the NOTICE or LICENSE file. This applies to this assembly NOTICE and the nar NOTICE.

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,433 @@
+                        listReader.read();
+                    }
+                    while(listReader.read(cellProcs) != null) {}
+                } catch (final IOException | SuperCsvCellProcessorException e) {
```
This should break IOException and SuperCsvCellProcessorException into separate catch blocks. A SuperCsvCellProcessorException should be logged at DEBUG because it will get thrown for invalid files as a normal part of processing for this processor. 

For example, I used your template with the schema "Null, ParseDate("dd/MM/yyyy"), Optional(ParseDouble())". I had the second column have a value of "22/111954". It failed to parse the data and correctly routed it to invalid but logged an ERROR bulletin.

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,433 @@
+                return null;
+
+            default:
+                throw new IllegalArgumentException(method + " is not an allowed method to define a Cell Processor");
```
This should quotes around "method". If a user has two commas in a row it can be quite confusing. 

Also an explicit check if "method" is empty (suggesting the user has two commas in a row) would be nice.

JPercivall commented:
I'm trying to use a couple of the processors listed in the accompanying doc you link to[1] but there seem to be some processors that aren't available (notably "IsIncludedIn"). I was trying to make sure a column had a value in a set of strings ("male" or "female"). 

Is there a reason for not including all the processors available?

[1] http://super-csv.github.io/super-csv/cell_processors.html
[2] http://super-csv.github.io/super-csv/apidocs/org/supercsv/cellprocessor/constraint/IsIncludedIn.html

pvillard31 commented:
Hey @JPercivall 
- Reverted NOTICE changes
- Separated the exception and changed the logging approach when there is an invalid CSV
- Changed the error message when the user provides a wrong method in the schema property and added a test (let me know in case it is not what you were thinking about)
- Added IsIncludedIn cell processor (my bad!).
- Added list of supported cell processors in additional documentation.

For the other cell processors I am not sure to see how they would fit in a validating process. If I am missing something here, let me know. There are interesting cell processors to modify CSV content by writing a new CSV file (a new FlowFile). This could be nice but this is not only validation anymore, I assume this would be better in a different processor, no?

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -408,9 +407,12 @@ public void process(final InputStream in) throws IOException {
                     logger.error("Failed to validate {} against schema due to {}", new Object[]{flowFile}, e);
+                } catch (final SuperCsvCellProcessorException e) {
+                    valid.set(false);
+                    logger.info("Failed to validate {} against schema due to {}; routing to 'invalid'", new Object[]{flowFile}, e);
```
This should be debug, not info, this will occur as a part of normal operation and is only really relevant for debugging.

JPercivall replied:
Actually some form of this should be made into an attribute to be added to invalid FlowFiles. That way users can know what is actually invalid.

JPercivall commented:
@pvillard31, sorry for not realizing this earlier but a great improvement that could be made is to route like RouteText does. Where it can either route the whole FlowFile when it fails or route each line to the respective valid/invalid destination. Where invalid rows would have an attribute detailing what went wrong.

I believe all that would need to be done is to create flowfiles with each reading of the row and if it fails add it to the invalid list (along with adding an attribute to it).

pvillard31 commented:
Hey @JPercivall, I finally found the time to address your comment. I agree that offering the possibility to filter out the "wrong" lines is a nice feature. I chose the option to have one flow file containing valid lines and one containing invalid lines. WRT this choice, I am not sure it makes sens to add an attribute with an error message since the error could be different for each invalid lines. Let me know your thoughts! Thanks!

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,613 @@
+                session.remove(flowFile);
+            } else if (okCount.get() != 0) {
+                if(header) {
+                    totalCount.set(totalCount.get() - 1);
```
This gets decremented here but, I may just be missing it, I don't see the Header add to the totalCount.

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,613 @@
+    public static final AllowableValue VALIDATE_LINES_INDIVIDUALLY = new AllowableValue(routeLinesIndividually, routeLinesIndividually,
+            "In case an error is found, the input CSV file will be split into two FlowFiles: one routed to the 'valid' "
+                    + "relationship containing all the correct lines and one routed to the 'invalid' relationship containing all "
+                    + "the incorrect lines. Take care if choosing this option while using Unique cell processors in schema definition.");
```
I like calling out the Unique processor here but it'd nice to have more explicit why and what happens if you do use it.

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,613 @@
+
+        if(isWholeFFValidation) {
+            if (valid.get()) {
+                logger.info("Successfully validated {} against schema; routing to 'valid'", new Object[]{flowFile});
```
This line is probably too common to be "info". Debug would match the other paths in this block too.

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,613 @@
+            "UniqueHashCode", "IsIncludedIn");
+
+    private static final String routeWholeFlowFile = "Route to 'valid' the whole FlowFile if the CSV is valid";
+    private static final String routeLinesIndividually = "Route to 'valid' a FlowFile containing the valid lines and"
```
These property names are a bit too long, they don't fit on the drop-down for the allowable values. Any way to shorten them?

![screen shot 2016-07-27 at 7 12 47 pm](https://cloud.githubusercontent.com/assets/11302527/17195852/6be2fc36-542e-11e6-9488-a5aa905a1873.png)

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,613 @@
+     * @param isFirstLine true if this is the first line we append
+     * @return String to append in the flow file
+     */
+    private byte[] print(List<?> list, CsvPreference csvPref, boolean isFirstLine) {
```
Not sure why but after being routed through the processor instead of '\n' the new line character the literals "\" and "n" were added. Below are two screen shots of the data below and after (only line routed to invalid)

![screen shot 2016-07-27 at 7 23 52 pm](https://cloud.githubusercontent.com/assets/11302527/17196060/e4267118-542f-11e6-8831-83f8389a615a.png)
![screen shot 2016-07-27 at 7 23 55 pm](https://cloud.githubusercontent.com/assets/11302527/17196059/e424d600-542f-11e6-8e35-5f1274912ef1.png)

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,613 @@
+                                }
+                            }
+
+                        } catch (final SuperCsvCellProcessorException e) {
```
This should probably also catch "SuperCsvException". I ran into one when the number of columns I had as my processor list didn't match the number passed.


![screen shot 2016-07-27 at 7 35 46 pm](https://cloud.githubusercontent.com/assets/11302527/17196264/5594b3c2-5431-11e6-8014-b8f0743aa107.png)

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -0,0 +1,613 @@
+                                }));
+
+                                if(isFirstLine.get()) {
+                                    isFirstLine.set(false);
```
Currently valid and invalid share the same "isFirstLine". So if the first row is valid and the second is invalid the invalid FF will be prepended with the new line character(s)

pvillard31 commented:
Hey @JPercivall, I think I've addressed most of your comments:
- changed log level
- fixed decrementation (was not related to header but rather to 'finally' in 'while' loop)
- strategy naming and description
- fixed first line handling and added a unit test for that and to confirm Unique() behavior when validating line by line
- changed exception catching

It remains what you observed when displaying content of flow files after being processed by the processor. I've reproduced your observation but I didn't find any explanation. By any chance, do you know if there are some specific encoding related to the UI display? (I remember some discussions regarding how is processed the carriage return (shift + enter instead of \n) when used in a property in some processors) If you haves ideas, let me know!

Thanks.

JPercivall commented:
@pvillard31, I know what the problem with the end-line characters is. When going from the UI to Java, the characters are escaped so that what you input is transferred over to Java as is. So when you type the characters "\" and "\n" into the UI the Java string will end up being those two characters *not* the interpreted value "\n".

There's been some discussion about it before and how we need to make some change but it hasn't been a top priority. For now what is done, is something like is done here[1]. Where the default value is escaped and then in the OnScheduled[2] or as a separate method[3] it is interpreted. 

[1] https://github.com/apache/nifi/blob/1373bf672586ba5ddcfa697c45c832ccc79425cb/nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/listen/AbstractListenEventBatchingProcessor.java#L61-L61
[2] https://github.com/apache/nifi/blob/1373bf672586ba5ddcfa697c45c832ccc79425cb/nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/listen/AbstractListenEventBatchingProcessor.java#L97-L97
[3] https://github.com/apache/nifi/blob/cd846c8d627efb2606f72b6af009358dec27be63/nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/put/AbstractPutEventProcessor.java#L566-L566

pvillard31 commented:
Thanks @JPercivall !
I did the modification in the ``OnSchedule`` method and it is now working as expected. I also took the liberty to squash my commits. Let me know if there is something else.

JPercivall commented:
+1

Visually verified code and did a contrib check build. In a standalone instance ran multiple different schemas through ValidateCSV testing each of the properties, all worked as expected. Thanks @pvillard31, I will merge it in.
